File: OCP/Delphi/aplicado/ICalculadorRegra.pas
unit ICalculadorRegra;

interface

type
  ICalculadorRegra = interface
  	function PodeCalcular(medida: String): Boolean;
  	function Calcular(item: TItemCarrinho): Double;
  end;
  
  TCalculadorRegraUnidade = class(TInterfacedObject, ICalculadorRegra)
    public
      function PodeCalcular(medida: String): Boolean;
      function Calcular(item: TItemCarrinho): Double;
  end;

  TCalculadorRegraPeso = class(TInterfacedObject, ICalculadorRegra)
    public
      function PodeCalcular(medida: String): Boolean;
      function Calcular(item: TItemCarrinho): Double;
  end;

  TCalculadorRegraEspecial = class(TInterfacedObject, ICalculadorRegra)
    public
      function PodeCalcular(medida: String): Boolean;
      function Calcular(item: TItemCarrinho): Double;
  end;

implementation

//TCalculadorRegraUnidade

function TCalculadorRegraUnidade.PodeCalcular(medida: String): Boolean;
begin
  Result := medida = 'unidade';
end;

function TCalculadorRegraUnidadeCalcular(item: TItemCarrinho): Double;
begin
  Result := item.Quantidade * 5;
end;

//TCalculadorRegraPeso

function TCalculadorRegraPeso.PodeCalcular(medida: String): Boolean;
begin
  Result := medida = 'peso';  
end;

function TCalculadorRegraPesoCalcular(item: TItemCarrinho): Double;
begin
  Result := item.Quantidade * 4 / 1000;
end;

//TCalculadorRegraEspecial

function TCalculadorRegraEspecial.PodeCalcular(medida: String): Boolean;
begin
  Result := medida = 'especial';
end;

function TCalculadorRegraEspecialCalcular(item: TItemCarrinho): Double;
var
	pacoteDeTres: Integer;
begin
  Result = 0.0;
  Inc(Result, item.Quantidade * 0.4);
  pacoteDeTres := item.Quantidade / 3;
  Result := Result - pacoteDeTres * 0.2;
end;

end.

File: OCP/C#/aplicado/ICalculadorRegra.cs
interface ICalculadorRegra
{
	bool PodeCalcular(string medida);
	decimal Calcular(ItemCarrinho item);
}

class CalculadorRegraUnidade : ICalculadorRegra
{
	bool PodeCalcular(string medida)
		=> medida == "unidade";
		
	decimal Calcular(ItemCarrinho item)
		=> item.Quantidade * 5;
}

class CalculadorRegraPeso : ICalculadorRegra
{
	bool PodeCalcular(string medida)
		=> medida == "peso";
		
	decimal Calcular(ItemCarrinho item)
		=> item.Quantidade * 4 / 1000;
}

class CalculadorRegraEspecial : ICalculadorRegra
{
	bool PodeCalcular(string medida)
		=> medida == "especial";
		
	decimal Calcular(ItemCarrinho item)
	{
		var total = item.Quantidade * 0.4;
  		var pacoteDeTres = item.Quantidade / 3;
  		return total - pacoteDeTres * 0.2;
	}
}